rmw_iceoryx_cpp/src/internal/iceoryx_name_conversion.cpp

```diff
@@ -93,20 +69,56 @@ get_name_n_type_from_iceoryx_service_description(
   const std::string & instance,
   const std::string & event)
 {
-  std::string topic_name;
-  std::string type_name;
-
-  if (service.find("/msg/") != std::string::npos) {
+  if (event == "data") {
     // ROS 2.0 Naming
     return std::make_tuple(instance, service);
-  } else {
-    // ARA Naming
-    std::string delimiter_msg = "_ara_msgs/msg/";
+  }
+  // ARA Naming
+  std::string service_lowercase = service;
+  std::transform(
+    service_lowercase.begin(), service_lowercase.end(),
+    service_lowercase.begin(), ::tolower);
+
+  return std::make_tuple(
+    "/" + instance + "/" + service + "/" + event,
+    service_lowercase + ARA_DELIMITER + event);
+}
 
-    return std::make_tuple(
-      "/" + instance + "/" + service + "/" + event,
-      service + delimiter_msg + event);
+std::tuple<std::string, std::string, std::string> get_service_description_elements(
+  const std::string & topic_name,
+  const std::string & type_name)
+{
+  auto position_ara_delimiter = type_name.find(ARA_DELIMITER);
+
+  if (position_ara_delimiter == std::string::npos) {
+    // ROS 2.0 Naming
+    return std::make_tuple(type_name, topic_name, "data");
+  }
+
+  // ARA Naming
+  // Due to ros package naming conventions the service name packed into
+  // the type name had to be lowercase
+  auto service_lowercase = type_name.substr(0, position_ara_delimiter);
+
+  std::string topic_name_lowercase = topic_name;
+  std::transform(
+    topic_name_lowercase.begin(), topic_name_lowercase.end(),
+    topic_name_lowercase.begin(), ::tolower);
+
+  // Find the lowercase service name in the lowercased topic name
+  auto position_package_name = topic_name_lowercase.find(service_lowercase);
+
+  if (position_package_name == std::string::npos) {
+    throw std::runtime_error("message topic and type are inconsistent");
   }
+
+  // Get the mixed uppercase and lowercase service name
+  // knowing the strings position in the topic name
+  auto service = topic_name.substr(position_package_name, service_lowercase.length());
+  auto instance = topic_name.substr(1, position_package_name - 2);
+  auto event = type_name.substr(position_ara_delimiter + strlen(ARA_DELIMITER), type_name.size());
```

Comment:
Wouldn't it be easier to get service, instance and event all from the topic name. This should include all three unchanged and separated by "/"? Or do I miss a fact?

Reply:
We could do this .. if there would be a convention to not put slashes in the instance-, service- or event-name. Since we have no convention in place I would like to stay generic to a certain extend.